src/coins.cpp

```diff
@@ -690,25 +724,54 @@ bool CCoinsViewCache::BatchWrite(CCoinsMap &mapCoins,
     }
     mapSidechainEvents.clear();
 
-    for (CCswNullifiersMap::iterator child_it = cswNullifiers.begin(); child_it != cswNullifiers.end();)
-    {
-        if (child_it->second.flags & CNullifiersCacheEntry::DIRTY) { // Ignore non-dirty entries (optimization).
-            CCswNullifiersMap::iterator parent_it = cswNullifiers.find(child_it->first);
+    for (auto& entryToWrite : cswNullifiers) {
+        CCswNullifiersMap::iterator itLocalCacheEntry = cacheCswNullifiers.find(entryToWrite.first);
 
-            if (parent_it == cswNullifiers.end()) {
-                CCswNullifiersCacheEntry& entry = cswNullifiers[child_it->first];
-                entry.entered = child_it->second.entered;
-                entry.flags = CCswNullifiersCacheEntry::DIRTY;
-            } else {
-                if (parent_it->second.entered != child_it->second.entered) {
-                    parent_it->second.entered = child_it->second.entered;
-                    parent_it->second.flags |= CCswNullifiersCacheEntry::DIRTY;
-                }
-            }
+        switch (entryToWrite.second.flags) {
+            case CCswNullifiersCacheEntry::FRESH:
+                assert(
+                    itLocalCacheEntry == cacheCswNullifiers.end() ||
+                    itLocalCacheEntry->second.flags == CCswNullifiersCacheEntry::ERASED
+                ); //A fresh entry should not exist in localCache or be already erased
+                cacheCswNullifiers[entryToWrite.first] = entryToWrite.second;
+                break;
+            case CCswNullifiersCacheEntry::ERASED:
+                if (itLocalCacheEntry != cacheCswNullifiers.end())
+                    itLocalCacheEntry->second.flags = CCswNullifiersCacheEntry::ERASED;
+                break;
+            default:
+                assert(false);
+        }
+    }
+    cswNullifiers.clear();
+
+    for (auto& entryToWrite : certDataHashes) {
+        CCertDataHashMap::iterator itLocalCacheEntry = cacheCertDataHashes.find(entryToWrite.first);
+
+        switch (entryToWrite.second.flags) {
+            case CCertDataHashCacheEntry::Flags::FRESH:
+                assert(
+                    itLocalCacheEntry == cacheCertDataHashes.end() ||
+                    itLocalCacheEntry->second.flags == CCertDataHashCacheEntry::Flags::ERASED
+                ); //A fresh entry should not exist in localCache or be already erased
+                cacheCertDataHashes[entryToWrite.first] = entryToWrite.second;
+                break;
+            case CCertDataHashCacheEntry::Flags::DIRTY: //A dirty entry may or may not exist in localCache
+                cacheCertDataHashes[entryToWrite.first] = entryToWrite.second;
```

Comment:
@abi87 What if in the cache some entry was marked as `FRESH`, but in the `entryToWrite` it's `DIRTY`?
So the entry was created initially in the first layer cache, and then modified in the second layer cache.
I think, that we need to leave in the `cacheCertDataHashes[entryToWrite.first]` flag value as is, ignoring the flag in the `entryToWrite.second`.
PS: though the current code in txdb processes `FRESH` and `DIRTY` in the same way, but we need to be more careful in the cache level.
What do you think?